indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -188,14 +194,36 @@ public OverlordResource(
   public Response taskPost(final Task task, @Context final HttpServletRequest req)
   {
     final String dataSource = task.getDataSource();
-    final ResourceAction resourceAction = new ResourceAction(
-        new Resource(dataSource, ResourceType.DATASOURCE),
-        Action.WRITE
-    );
+    final Set<ResourceAction> resourceActions = new HashSet<>();
+    resourceActions.add(new ResourceAction(new Resource(dataSource, ResourceType.DATASOURCE), Action.WRITE));
+    if (authConfig.isEnableInputSourceSecurity()) {
+      if (task.usesFirehose()) {
+        return Response.status(Response.Status.BAD_REQUEST)
+            .entity(
+                ImmutableMap.of(
+                    "error",
+                    StringUtils.format(
+                        "Input source based security cannot be performed for Task[%s] because it uses firehose."
+                        + "Change the tasks configuration, or disable `isEnableInputSourceSecurity`",
+                        task.getId()
+                    )
+                )
+            )
+            .build();
+      }
+      if (null != task.getInputSourceTypes()) {
```

Comment:
Nit: we're computing the input source types twice.

```
Set<String> inputSourceTypes = task.getInputSourceTypes();
if (inputSourceTypes != null) {
...
```

Then, the reason we return a `Set` and not a single key is that some input sources are compound. If we allow the input source type method to return null, then those methods that are compound have to include code to handle null.

But, if we say the method must return a non-null set, then it is easy to combine child input sources without null checks.

Reply:
fixed